Record position precision

scd31 opened the pull request:
I started working on Mic-E encoding/decoding and realized it also deals with position precision. I figured it makes sense to add this now so that we can work with it in Mic-E.

@Turbo87 Tagging you as you requested this. Works in both directions (encoding and decoding)

scd31 commented on src/lonlat.rs:
```diff
@@ -42,50 +48,69 @@ impl Latitude {
+        total_spaces += num_spaces;
 
-        let (min_frac, _) = parse_bytes_trailing_spaces(&[b[5], b[6]], only_spaces)
+        let precision = Precision::from_num_digits(total_spaces)
```
The precision is only specified in the latitude but it applies to the longitude as well. I'm not sure if we should ignore the extra digits in the longitude or not. For Mic-E (which this isn't) the docs say we should. It doesn't say anything about the normal uncompressed format.

Ex. If we get `1234.  N5678.90W` should be decode the longitude as `5678.90` or `5678.00`? @benjamin-l any thoughts?

scd31 replied:
One could even make the argument it should be interpreted as `5678.50` since that lies right in the middle of the possible values

olivia-fl replied:
I think this really indicates an issue with our data representation. From page 34 of the spec:

> The level of ambiguity specified in the latitude will automatically apply to
the longitude as well — it is not necessary to include any `⎵` characters in the
longitude.
> 
> For example, the coordinates:
> `4903.⎵⎵N/07201.75W-`
> represent the position to the nearest minute. That is, the hundredths of
> minutes of latitude and longitude may take any value in the range 00–99.
>
> Thus the station may be located anywhere inside a bounding box having the
> following corner coordinates:
>  - North West corner: 49 deg 3.99 mins N, 72 deg 1.99 mins W
>  - North East corner: 49 deg 3.99 mins N, 72 deg 1.00 mins W
>  - South East corner: 49 deg 3.00 mins N, 72 deg 1.00 mins W
>  - South West corner: 49 deg 3.00 mins N, 72 deg 1.99 mins W

So `1234. N5678.90W` should be semantically identical to `1234. N5678.00W`. Our current representation treats the exact value parsed from the packet and the precision as separate objects with independent meaning, which is why we're having to ask questions about what exact value to choose in this situation in the first place.

I think the spec implies semantics more like this:

```rust
impl Precision {
    // The std range types are kinda painful to use. This could be replaced with
    // a tuple if we want to.
    fn range(value: f32) -> RangeInclusive<f32> {
        // left as an exercise to the reader :)
    }
}

struct AmbiguousValue {
    value: f32,
    pub precision: Precision
}

impl AmbiguousValue {
    fn range(&self) -> RangeInclusive<f32> {
        self.precision.range(self.value)
    }

    fn center(&self) -> f32 {
        let (low, high) = self.range().into_inner();
        0.5 * high + 0.5 * low
    }

    /// Docs that explain why this _probably_ isn't what you want to do...
    ///
    /// In particular, you usually would want `center()` or `range().start()`
    fn raw_value(&self) -> f32 {
        self.value
    }
}

struct Latitude(pub AmbiguousValue);
struct Longitude(pub AmbiguousValue);

// or maybe even this, to reduce the invalid state space further
struct LonLat {
    pub precision: f32,
    latitude: f32,
    longitude: f32,
}

impl LonLat {
    fn latitude(&self) -> AmbiguousValue {
        AmbiguousValue {
            value: self.latitude,
            precision: self.precision
        }
    }

    fn longitude(&self) -> AmbiguousValue {
        AmbiguousValue {
            value: self.longitude,
            precision: self.precision
        }
    }
}
```

Unfortunately this complicates the API quite a bit... If we're going to change the position representation this drastically, it might be worth also thinking about swapping out the floats for fixed point (integer number of 1/100 minute), and then providing a conversion to float.

Either way, whether we go for this kind of representation or keep the current representation, I think it should be decoded as `5678.90`. The fact that the last two digits aren't semantically meaningful is the responsibility of whatever's interpreting the data, not the parser.

scd31 replied:
> Either way, whether we go for this kind of representation or keep the current representation, I think it should be decoded as 5678.90. The fact that the last two digits aren't semantically meaningful is the responsibility of whatever's interpreting the data, not the parser.

Hmm, not sure if I agree with this. If the last two digits are meaningless then I think we would want to get rid of them entirely. I'm not even necessarily sure if they would have to be digits. For Mic-E encoding, the excess longitude digits don't need to be valid, if I'm reading the spec correctly, so I think it might make sense to be consistent and ignore it for the normal uncompressed positions too. I'm definitely open to other thoughts on this though.

scd31 replied:
Another thing we should probably think about is how we want the user to construct positions with ambiguity. We would probably want them to specify lat/lon together along with a precision value (to prevent them from specifying a different precision for each). The problem with the precision is that it's pretty specific to the internal representation of APRS location so it would be good to abstract it away. I'm not sure if this is possible though.

scd31 replied:
What about something like this?

```rust
// generic so that we can reuse it for MIC-E
trait Precision {
    fn range();
    fn center(); // this could be implemented in the trait itself using range()
    // allows finding the closest precision to a given range. Hopefully useful for constructing packets
    fn closest_precision(range: RangeInclusive) -> Self;
}

enum UncompressedPrecision {
    // all the variants in the current Precision that's in this PR
}

impl Precision for UncompressedPrecision {
    // ...
}

struct AmbiguousValue<P: Precision> {
    value: f32,
    precision: P
}

struct AprsPosition {
    latitude: Latitude,
    longitude: Longitude,
    precision: UncompressedPrecision
}

impl AprsPosition {
    fn new(latitude: Latitude, longitude: Longitude, precision: UncompressedPrecision) -> Self {
        // left as an exercise to the reader
    }

    fn latitude(&self) -> AmbiguousValue<UncompressedPrecision> {
        AmbiguousValue {
            value: self.latitude,
            precision: self.precision
        }
    }

    fn longitude(&self) -> AmbiguousValue<UncompressedPrecision> {
        AmbiguousValue {
            value: self.longitude,
            precision: self.precision
        }
    }
}
```

I do have to walk back a bit of what I said. The user would have to have some understanding of how positions work internally in APRS. Also, the `UncompressedPrecision` (and future `MicEPrecision`) would be exposed to the users. I don't think it's that big of a deal though.

Edit: Also - I didn't bother making the change in this code example but I'm heavily in favour of using fixed-precision integers instead of floats for the values

olivia-fl replied:
> I'm not even necessarily sure if they would have to be digits

Oh shit I did not realize that. If this is the case I think it definitely makes sense to strip them.

If we are going to strip them, I think the best way to do it is probably truncating rather than trying to do any fancy rounding math. It doesn't really matter as long as the described range is the same.

For Mic-E, are the precision semantics actually different? I skimmed the spec and it looked like it was the same kind of thing with "ignore the last N digits"...

scd31 replied:
I took a closer look and you're right, Mic-E seems to have the same precision semantics. We probably just need the one precision type.

scd31 commented on src/lonlat.rs:
```diff
@@ -42,50 +48,69 @@ impl Latitude {
         let min = ((lat - (deg as f64)) * 60.0) as u32;
         let min_frac = ((lat - (deg as f64) - (min as f64 / 60.0)) * 6000.0).round() as u32;
 
-        write!(buf, "{:02}{:02}.{:02}{}", deg, min, min_frac, dir)?;
+        // into_bytes is safe because we know the string only contains ASCII values
+        let mut digit_buffer = format!("{:02}{:02}{:02}", deg, min, min_frac).into_bytes();
+
+        // zero out fields as required for precision
+        // Ideally we would be doing some clever rounding here
+        // E.g. if last 2 digits were blanked,
+        // 4905.83 would become 4906.__
+        for i in (6 - precision.num_digits())..6 {
+            digit_buffer[i as usize] = b' ';
+        }
+
+        buf.write_all(&digit_buffer[0..4])?;
+        write!(buf, ".")?;
+        buf.write_all(&digit_buffer[4..6])?;
+        write!(buf, "{}", dir)?;
         Ok(())
```
I truly hate this. It should work fine but it looks bad, and will need to be rewritten anyway for no-std (I plan on doing this eventually so that this library can be used in embedded systems)

olivia-fl replied:
Yeah... I don't have a better idea yet, but will think about it.

olivia-fl replied:
Here's an option that still doesn't do any rounding, but is maybe simpler than the current code:

```rust
let mut digit_buffer = [b' '; 6];
let blank_index = 6 - precision.num_digits();
write!(&mut digit_buffer[..blank_index], "{:02}{:02}{:02}", deg, min, min_frac).unwrap();
buf.write_all(&digit_buffer[0..4])?;
write!(buf, ".")?;
buf.write_all(&digit_buffer[4..6])?;
write!(buf, "{}", dir)?;
```

Also still not `no_std` compatible, because it uses `Write`, but that's going to be a much larger issue :)

olivia-fl commented on src/position.rs:
```diff
@@ -30,6 +31,49 @@ pub enum AprsCst {
+}
+
+impl Precision {
+    pub(crate) fn num_digits(&self) -> u8 {
```
These functions seem pretty useful to downstream users, might want them to be `pub` instead of `pub(crate)`.

scd31 replied:
My only concern is that the # of digits seems fairly specific to APRS. The first two digits are degrees and thus have a different significance than the other digits.

If we went the `AmbiguousValue` route (which I am heavily in favour of) I would be tempted to make `Precision` an internal-only type

olivia-fl replied:
I think even in the `AmbiguousValue` version, I still think it makes sense to expose what `Precision` actually means to users. Yes, it's APRS specific, but this is an APRS encoding library.

I'm somewhat on board with hiding `num_digits` though, since it's specific to the `deg min.min/100` representation, which isn't something that's really exposed elsewhere.

scd31 commented:
@Benjamin-L I've pushed up some changes.

- The API is different from what we described. I kept the latitude/longitude as-is and added methods for getting their bounds (using the precision). I think this is best as most people are just going to want the actual values 99% of the time, and anyone who actually cares about precision can just call the methods manually. This also solves the problem of instantiating an `AprsPosition` (with the `AmbiguousValue` type it was going to be a bit complicated). Let me know what you think

- I updated the longitude parser to ignore any characters that the precision says we can ignore about

- I updated the logic for encoding precisions from your suggestions

olivia-fl commented:
@scd31 the new version looks good to me
